### PR TITLE
feat(auth): HostedUI oidc signout

### PR DESCRIPTION
**Pull request by Samaritan1011001**
#### Description of changes

Updates the signOut API to allow customers to provide which redirect url to use enabling sign out of users from both Cognito Hosted UI and the upstream OIDC provider.

#### Description of how you validated changes
Sample app with HostedUI setup with Auth0 as the OIDC provider.
- sign in using signWithRedirect
- sign out with preferredRedirectUrl as the Auth0's logout endpoint
- call signInWithRedirect to see that the Auth0 login again requires to enter user credentials.

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts [hunk 1]
```diff
@@ -6,18 +6,33 @@ import {
...
-		redirects?.find(isTheSameDomain);
-	const redirectUrlFromDifferentOrigin =
-		redirects?.find(isHttps) ?? redirects?.find(isHttp);
-	if (redirectUrlFromTheSameOrigin) {
-		return redirectUrlFromTheSameOrigin;
-	} else if (redirectUrlFromDifferentOrigin) {
-		throw invalidOriginException;
+export function getRedirectUrl(
+	redirects: string[],
+	_preferredSignOutUrl?: string,
```
Comment (cshfang) [inline packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts:11]: What's the reason for adding the `_`?

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts [hunk 2]
```diff
@@ -6,18 +6,33 @@ import {
...
+	if (_preferredSignOutUrl) {
+		const redirectUrl = redirects?.find(
+			redirect => redirect === _preferredSignOutUrl,
+		);
+		if (!redirectUrl) {
+			throw invalidRedirectException;
+		}
+
+		return redirectUrl;
+	} else {
```
Comment (cshfang) [inline packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts:22]: Nit: If clause above has early return, so nesting the rest in the `else` block should be unnecessary

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts [hunk 3]
```diff
@@ -4,14 +4,27 @@
...
 		throw invalidRedirectException;
 	}
+	if (preferredSignOutUrl) {
+		preferredRedirectUrl = redirects?.find(
+			redirect => redirect === preferredSignOutUrl,
+		);
+		if (!preferredRedirectUrl) {
+			throw invalidRedirectException;
+		}
+	}
```
Comment (cshfang) [inline packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts:27]: Couldn't this have used the exact same strategy as the web version?
```ts
if (preferredSignOutUrl) {
	const redirectUrl = redirects?.find(
		redirect => redirect === _preferredSignOutUrl,
	);

	if (!redirectUrl) {
		throw invalidRedirectException;
	}

	return redirectUrl;
}

// existing previous logic
```

**Review by israx — COMMENTED** [review]
Thanks Manoj. Left a couple of comments

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts (hunk 1 above)
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts:11]: I want to say the linter was complaining but not able to repro it so I removed the _

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts (hunk 2 above)
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts:22]: Yes but I feel its better for readability with the else block there. It's clear which of the two scenarios to check when reading it.

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts (hunk 3 above)
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts:27]: Yes it can be the same, but in the mobile case we should fail early if there is no app scheme found in the array. It does matter what the preferred url is if there is no app scheme in the array for us to register. In the web version, the preferred url takes precedence.

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts [hunk 4]
```diff
@@ -4,14 +4,27 @@
...
-	if (!redirectUrl) {
+	if (!appSchemeRedirectUrl) {
 		throw invalidRedirectException;
 	}
+	if (preferredSignOutUrl) {
+		preferredRedirectUrl = redirects?.find(
+			redirect => redirect === preferredSignOutUrl,
+		);
+		if (!preferredRedirectUrl) {
+			throw invalidRedirectException;
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts:25]: That's fair. Updated.

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts [hunk 5]
```diff
@@ -2,22 +2,38 @@
...
+	if (preferredSignOutUrl) {
+		const redirectUrl = redirects?.find(
+			redirect => redirect === preferredSignOutUrl,
+		);
+		if (!redirectUrl) {
+			throw invalidPreferredRedirectUrlException;
+		}
+
+		return redirectUrl;
+	} else {
```
Comment (HuiSF) [inline packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts:23]: nit: you meant the following?

```suggestion
	if (preferredSignOutUrl) {
		if (redirects?.includes(preferredSignOutUrl)) {
			return redirectUrl;
		}

		throw invalidPreferredRedirectUrlException;
	} else {
```

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts [hunk 6]
```diff
@@ -1,17 +1,38 @@
...
+	const appSchemeRedirectUrl = redirects?.find(
 		redirect =>
 			!redirect.startsWith('http://') && !redirect.startsWith('https://'),
 	);
-	if (!redirectUrl) {
-		throw invalidRedirectException;
+	if (!appSchemeRedirectUrl) {
+		throw invalidAppSchemeException;
 	}
+	if (preferredSignOutUrl) {
```
Comment (HuiSF) [inline packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts:26]: nit: could move this to the top to avoid executing the look up of `appSchemeRedirectUrl`?

File: packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.test.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	beforeEach(() => {
+		windowSpy = jest.spyOn(window, 'window', 'get');
+	});
+
+	afterEach(() => {
+		windowSpy.mockRestore();
+	});
+
+	it('should return the redirect url that has the same origin and same pathName', () => {
+		windowSpy.mockImplementation(() => ({
```
Comment (HuiSF) [inline packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.test.ts:21]: nit: could use `mockReturnValueOnce`

File: packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.test.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+			location: {
+				origin: 'https://example.com/',
+				pathname: 'app',
+			},
+		}));
+		expect(getRedirectUrl(mockRedirectUrls)).toStrictEqual(mockRedirectUrls[0]);
+	});
+
+	it('should throw an invalid origin exception if there is no url that is the same origin and pathname', () => {
+		windowSpy.mockImplementation(() => ({
```
Comment (HuiSF) [inline packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.test.ts:31]: nit: ditto

File: packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.test.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+			},
+		}));
+		expect(() => getRedirectUrl(mockRedirectUrls)).toThrow(
+			invalidOriginException,
+		);
+	});
+
+	it('should throw an invalid redirect exception if there is no url that is the same origin/pathname and is also not http or https', () => {
+		const mockNonHttpRedirectUrls = ['test-non-http-string'];
+		windowSpy.mockImplementation(() => ({
```
Comment (HuiSF) [inline packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.test.ts:44]: nit: ditto

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts (hunk 5 above)
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.ts:23]: That reads well, thanks, I'll update it.

File: packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts (hunk 6 above)
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/utils/oauth/getRedirectUrl.native.ts:26]: We need `appSchemeRedirectUrl` irrespective of the `preferredSignOutUrl`. We don't return the appScheme above if you notice. Flow is like, throw `appSchemeNotFound` error since that check is always required then check for the `preferredSignOutUrl` and return it if present, if not proceed with the appScheme.

File: package.json [hunk 10]
```diff
@@ -137,5 +137,8 @@
 		"nx": "16.7.0",
 		"xml2js": "0.5.0",
 		"tar": "6.2.1"
+	},
+	"overrides": {
+		"tar": "6.2.1"
```
Comment (israx) [inline package.json:142]: is there any reason to add a resolution at the root here?

Comment (Samaritan1011001) [reply package.json:142]: So, this is required to override the version when using `npm` because `resolutions` block doesn't work with `npm`. 
And this version is the secure version of `tar` that needs to be used. Found via `npm audit`

File: .github/workflows/callable-e2e-test-detox.yml [hunk 11]
```diff
@@ -13,11 +13,15 @@ on:
...
+      host_signout_page:
+        required: false
+        type: boolean
+        default: false
 
 jobs:
   e2e-test:
     name: E2E-Detox ${{ inputs.test_name }}
-    runs-on: macos-latest
+    runs-on: macos-latest-large
```
Comment (Samaritan1011001) [inline .github/workflows/callable-e2e-test-detox.yml:24]: Using this large image allows the detox to load the RN app on startup. Without, tests take atleast 2 trys to succeed.

File: .github/workflows/callable-e2e-test-detox.yml [hunk 12]
```diff
@@ -70,6 +74,14 @@ jobs:
           JEST_JUNIT_OUTPUT_NAME: detox-test-results.xml
         working-directory: ${{ inputs.working_directory }}
         shell: bash
+      - name: Start the http-server and host the oidc signout page locally (background)
```
Comment (Samaritan1011001) [inline .github/workflows/callable-e2e-test-detox.yml:77]: Step to boot the localhost server.

File: packages/auth/src/types/inputs.ts [hunk 13]
```diff
@@ -46,7 +46,10 @@ export interface AuthSignInInput<
 	options?: ServiceOptions;
 }
 export interface AuthSignOutInput {
-	global: boolean;
+	global?: boolean;
+	oauth?: {
```
Comment (Samaritan1011001) [inline packages/auth/src/types/inputs.ts:50]: API signature changes

File: packages/rtn-web-browser/package.json [hunk 14]
```diff
@@ -25,7 +25,7 @@
 	},
 	"devDependencies": {
 		"@types/react-native": "0.70.0",
-		"react-native": "0.72.3",
+		"react-native": "0.72.15",
```
Comment (Samaritan1011001) [inline packages/rtn-web-browser/package.json:28]: Needed to close a vulnerability on the bad version.

File: .github/workflows/callable-e2e-test-detox.yml [hunk 15]
```diff
@@ -70,6 +74,14 @@ jobs:
...
+        if: ${{ inputs.host_signout_page }}
+        run: |
+          cd oidc-signout-hosted-page/ &&
+          npx http-server -p 8000 &
```
Comment (israx) [inline .github/workflows/callable-e2e-test-detox.yml:81]: is this command going to run at the root of the RN app or the monorepo ?

**Review by israx — DISMISSED** [review]

File: packages/auth/src/types/inputs.ts [hunk 16]
```diff
@@ -46,7 +46,10 @@ export interface AuthSignInInput<
 	options?: ServiceOptions;
 }
 export interface AuthSignOutInput {
-	global: boolean;
+	global?: boolean;
```
Comment (cshfang) [inline packages/auth/src/types/inputs.ts:49]: Is this possibly a breaking change
```ts
const doSomethingIfGlobal = (global: boolean) => {
  if (global) {
    // do something
  }
}

const mySignOut = (input: AuthSignOutInput) => {
  doSomethingIfGlobal(input.global);
}
```

Comment (israx) [reply packages/auth/src/types/inputs.ts:49]: Yeah good catch. I guess we have a trade off between a possible breaking change for an uncommon use case, vs forcing the usage of a parameter that is not needed.

Comment (israx) [reply packages/auth/src/types/inputs.ts:49]: I would rather avoid forcing the usage of the `global` flag to prevent confusion and potential impact on the UX for authenticated users if passed in without knowing what it does.

Comment (Samaritan1011001) [reply packages/auth/src/types/inputs.ts:49]: Yeah, the reasoning behind making it optional is this. 
`If the global parameter is NOT optional, customers will have to provide a value for global if they want to provide a preferredRedirectUrl.`

File: .github/workflows/callable-e2e-test-detox.yml (hunk 15 above)
Comment (HuiSF) [reply .github/workflows/callable-e2e-test-detox.yml:81]: Is it possible to encapsulate the http server bootstrap command into a package.json script under the RN test app?

File: package.json (hunk 10 above)
Comment (HuiSF) [reply package.json:143]: I think we kept only resolutions based on the assumption that `yarn` is the driver here. If we want to also use npm, should we create a 100% mapping from `resolutions`?

File: packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.native.test.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+	it('should return preferredRedirectUrl if it matches at least one of the redirect urls from config', () => {
+		const configRedirectUrl = mockRedirectUrls[2];
+
+		expect(getRedirectUrl(mockRedirectUrls, configRedirectUrl)).toStrictEqual(
+			configRedirectUrl,
+		);
+	});
+
+	it('should throw an exception when there is no url with no http or https as prefix irrespective if a preferredSignOutUrl is given or not', () => {
```
Comment (HuiSF) [inline packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.native.test.ts:23]: sanity check: this description sounds like that `getRedirectUrl.native.test.ts` always expects at least one custom app scheme, is this correct? Is a universal link (starts with `https`) supposed to work, too?

File: packages/auth/src/providers/cognito/utils/oauth/oAuthSignOutRedirect.ts [hunk 18]
```diff
@@ -12,11 +12,12 @@ import { getRedirectUrl } from './getRedirectUrl';
 export const oAuthSignOutRedirect = async (
 	authConfig: CognitoUserPoolConfig,
 	preferPrivateSession = false,
+	preferredSignOutUrl?: string,
```
Comment (HuiSF) [inline packages/auth/src/providers/cognito/utils/oauth/oAuthSignOutRedirect.ts:15]: nit: can we use named parameter here?

```
{
  authConfig: CognitoUserPoolConfig,
  preferPrivateSession?: boolean,
  preferredSignOutUrl: string | undefined // required param but allow undefined
}
```

Concern: `preferredSignOutUrl` is currently a optional and positional parameter, calling `oAuthSignOutRedirect` without passing anything as the 3rd param is totally fine. With this, for example, while refactoring someone may forget to pass this param, and can't catch it early.

File: packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.native.test.ts (hunk 17 above)
Comment (Samaritan1011001) [reply packages/auth/__tests__/providers/cognito/utils/oauth/getRedirectUrl.native.test.ts:23]: Yeah, it always expect at-least one custom app scheme, it is the current behavior and we are not changing that with this change. We don't support universal links yet.

File: .github/workflows/push-integ-test.yml [hunk 19]
```diff
@@ -8,7 +8,7 @@ concurrency:
 on:
   push:
     branches:
-      - replace-with-your-branch
+      - fix/redirect-rn
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply .github/workflows/push-integ-test.yml:11]: It is intentional but I will revert it soon.

File: .github/workflows/callable-e2e-test-detox.yml (hunk 15 above)
Comment (Samaritan1011001) [reply .github/workflows/callable-e2e-test-detox.yml:81]: Root of the RN app. 

Hui, yeah, I thought of it initially, should be possible.

File: package.json (hunk 10 above)
Comment (Samaritan1011001) [reply package.json:143]: I would lean towards adding the other `overrides` as the need arises. `tar` version is the only one showing up so far for `npm audit`.

File: .github/workflows/callable-e2e-test-detox.yml (hunk 11 above)
Comment (ashika112) [reply .github/workflows/callable-e2e-test-detox.yml:24]: And the CI workflow works as expected?

Comment (Samaritan1011001) [reply .github/workflows/callable-e2e-test-detox.yml:24]: Yes, it does.

File: .github/workflows/push-integ-test.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply .github/workflows/push-integ-test.yml]: Reverted

File: .github/workflows/callable-e2e-test-detox.yml (hunk 15 above)
Comment (Samaritan1011001) [reply .github/workflows/callable-e2e-test-detox.yml]: @HuiSF Done.

File: .github/workflows/push-integ-test.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (israx) [reply .github/workflows/push-integ-test.yml:11]: does it need to be reverted again?

[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply .github/workflows/push-integ-test.yml:11]: Yes, I will remove this once this [PR](https://github.com/aws-amplify/amplify-js-samples-staging/pull/949/files) gets merged.

File: .github/workflows/push-preid-release.yml [hunk 20]
```diff
@@ -34,5 +34,4 @@ jobs:
     uses: ./.github/workflows/callable-npm-publish-preid.yml
     # The preid should be detected from the branch name recommending feat/{PREID}/whatever as branch naming pattern
     #   if your branch doesn't follow this pattern, you can override it here for your branch.
-    with:
-      preid: ${{ needs.parse-preid.outputs.preid }}
+    with: preid:${{ needs.parse-preid.outputs.preid }}
```
Comment (HuiSF) [inline .github/workflows/push-preid-release.yml:37]: Is this change intentional?

File: packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.native.ts [hunk 21]
```diff
@@ -5,20 +5,24 @@ import { CognitoUserPoolConfig } from '@aws-amplify/core';
...
 import { DefaultOAuthStore } from '../../utils/signInWithRedirectStore';
+import { TokenOrchestrator } from '../../tokenProvider';
 
 import { completeOAuthSignOut } from './completeOAuthSignOut';
 import { oAuthSignOutRedirect } from './oAuthSignOutRedirect';
 
 export const handleOAuthSignOut = async (
 	cognitoConfig: CognitoUserPoolConfig,
 	store: DefaultOAuthStore,
+	// No-op here as it's only used in the non-native implementation
```
Comment (HuiSF) [inline packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.native.ts:16]: nit: if we covert the positional params into named params then we don't need to pass in this?

File: .github/workflows/push-preid-release.yml (hunk 20 above)
Comment (Samaritan1011001) [reply .github/workflows/push-preid-release.yml:37]: No, reverted.

File: packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.native.ts (hunk 21 above)
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.native.ts:16]: Discussed offline, will keep it the same to reduce complexity on the changes.

**Review by HuiSF — DISMISSED** [review]

**Review by israx — DISMISSED** [review]

**Review by cshfang — APPROVED** [review]

File: packages/auth/src/providers/cognito/apis/signOut.ts [hunk 22]
```diff
@@ -60,21 +60,20 @@ export async function signOut(input?: SignOutInput): Promise<void> {
...
 				oAuthStore,
 				tokenOrchestrator,
+				input?.oauth?.preferredSignOutUrl,
 			)) ?? {};
 		if (type === 'error') {
 			throw new AuthError({
 				name: OAUTH_SIGNOUT_EXCEPTION,
-				message:
-					'An error occurred when attempting to log out from OAuth provider.',
+				message: `An error occurred when attempting to log out from OAuth provider. ${error}`,
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/auth/src/providers/cognito/apis/signOut.ts]: This diff seems to be outdated, exposing the error is not in main so I left it as is.

File: .github/workflows/push-integ-test.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply .github/workflows/push-integ-test.yml]: Merged that PR, so reverted this.

**Review by jimblanc — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]
